### PR TITLE
ui: serve ETag header and respect If-None-Match req. header for assets

### LINKED ISSUE #62708: Implement caching for Admin UI JS
**Is your feature request related to a problem? Please describe.**
Loading the admin UI is frequently very slow for me. This is because the files are served with no-cache headers, which means that the browser will frequently have to download the entire JS bundle:
Screen Shot 2021-03-29 at 6 25 38 PM

The connection to the cockroachdb admin from my local machine isn't super fast, however even on fast connections loading 25MB of JS is not going to be speedy.

**Describe the solution you'd like**
Pass cache-control headers which allow for long lived caching of the file, and cache bust using a query param or new filename when using a new version of cockroachdb. Also, why does the admin panel need 25MB of javascript - from a quick look at the generated bundles it looks like there's a lot of code which may not be needed.

**Additional context**
We're an enterprise customer.

Epic CC-5280

Jira issue: CRDB-2645

### PR DESCRIPTION
**Pull request by sjbarag**
CockroachDB has included a "Cache-Control: no-cache" response header when serving static, embedded assets for several years now (dating back to as early as 2015), but offered no other way for clients to cache those. Since then, asset sizes have grown considerably larger (currently including ~30MB of JavaScript, a size that will be addressed via other methods) and preventing client-side caching causes noticeable loading delays. Add support for the If-None-Match / ETag pair of request/response headers [1], which allows browsers serve assets from their caches if the cached copy of a requested asset is still the latest copy.

[1] https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match

fixes #62708

### CODE REVIEW

File: pkg/ui/buildutil/hash_files.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/cockroachdb/errors"
+)
+
+// HashFilesInDir recursively computes the SHA1 hash of every file in fsys
+// starting at root, and stores the computed hashes in dest["/path/to/file"]
+// (*including* a leading "/").
+func HashFilesInDir(dest *map[string]string, fsys fs.FS) error {
+	if dest == nil {
+		return errors.New("Unable to hash files without a hash destination")
+	}
```
Comment (sjbarag) [inline pkg/ui/buildutil/hash_files.go:28]: For some reason, github shows me a few files in this diff as being indented by one or two characters, but this definitely uses tab characters when I `cat` them locally (or when viewed through reviewable).  I've no idea what's causing that.

image

**Review by dhartunian — COMMENTED** [review]
> Reviewed 7 of 9 files at r1, 7 of 7 files at r3, 5 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78460)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @irfansharif, and @sjbarag)

___
*[pkg/ui/buildutil/hash_files.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-MyyIpsUlR0EfvhmYZTC-r2-28:-Mz1iGooEcLPfZEEIPL2:b-jft4sr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68da87e1a54b283615b82f0c32398e0af8f4ba7/pkg/ui/buildutil/hash_files.go#L28)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

For some reason, github shows me a few files in this diff as being indented by one or two characters, but this definitely uses tab characters when I `cat` them locally (or when viewed through reviewable).  I've no idea what's causing that or if you'll see it too.

image

</blockquote></details>

as long as you pass crlfmt lint, you should be good :)
___
*[pkg/ui/buildutil/hash_files.go, line 56 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1jnL-BDzzQakP79d4:-Mz1jnL-BDzzQakP79d5:b-qsb90s) ([raw file](https://github.com/cockroachdb/cockroach/blob/431671b927001264a34742ceb4bf309544f9782b/pkg/ui/buildutil/hash_files.go#L56)):*
> ```Go
> 
> 		// Store the computed hash
> 		fileHashes["/"+path] = hex.EncodeToString(hash.Sum(nil))
> ```

Q: I'm curious why the `"/"` is necessary
___
*[pkg/util/httputil/handlers.go, line 30 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1kNld0MIaRCABQfq_:-Mz1kNld0MIaRCABQfqa:b-vpxwpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3b1fed88cca24493a780140a530b9bf020f912b/pkg/util/httputil/handlers.go#L30)):*
> ```Go
> //	 value to use for that file
> // - next is the next handler in the http.Handler chain, used
> func EtagHandler(contentHashes *map[string]string, next http.Handler) http.Handler {
> ```

maps are already pointers so there's no need to use `*` to save your self a copy step if that's what you were worried about.
___
*[pkg/util/httputil/handlers.go, line 63 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1mADO2NlX3ZgioqSd:-Mz1mADO2NlX3ZgioqSe:bovdk33) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3b1fed88cca24493a780140a530b9bf020f912b/pkg/util/httputil/handlers.go#L63)):*
> ```Go
> 			if _, err := w.Write(nil); err != nil {
> 				log.Errorf(r.Context(), "Unable to write empty response body: %+v", err)
> 			}
> ```

nit: stylewise, we would typically return here and have the else clause on its own as the last line in the function.
___
*[pkg/util/httputil/etag_handler_test.go, line 54 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1meBhEJtBMyMk_a6Q:-Mz1meBhEJtBMyMk_a6R:b7ahj3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3b1fed88cca24493a780140a530b9bf020f912b/pkg/util/httputil/etag_handler_test.go#L54)):*
> ```Go
> 
> 	handler := EtagHandler(&hashedFiles, okHandler)
> 	server := httptest.NewServer(handler)
> ```

as discussed on Slack, probably need to close the server.

**Review by sjbarag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78460)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @irfansharif, and @sjbarag)

___
*[pkg/ui/buildutil/hash_files.go, line 56 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1jnL-BDzzQakP79d4:-MzGU8933F_1ae58wgh2:b5bq6x3) ([raw file](https://github.com/cockroachdb/cockroach/blob/431671b927001264a34742ceb4bf309544f9782b/pkg/ui/buildutil/hash_files.go#L56)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Q: I'm curious why the `"/"` is necessary
</blockquote></details>

`fs.WalkDir` passes a relative path to its callback as the `d` param, so nested assets would end up in `fileHashes` as `foo/bar/baz.js`.  The path provided in a request's `r.URL.Path` has a leading `/` though!  It seemed reasonable to include it here, but there's definitely some incongruity.

Think I should I rename `fileHashes` to `pathHashes` and keep the leading `/` here?  Or should I maybe remove the leading `/` from the `r.URL.Path` in the `ETag` handler?
___
*[pkg/util/httputil/handlers.go, line 30 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1kNld0MIaRCABQfq_:-MzGVECM6VT4W5YYtoTu:b-z7latu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3b1fed88cca24493a780140a530b9bf020f912b/pkg/util/httputil/handlers.go#L30)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

maps are already pointers so there's no need to use `*` to save your self a copy step if that's what you were worried about.
</blockquote></details>

That's exactly what I was worried about!  I actually had no idea — how handy!
___
*[pkg/util/httputil/handlers.go, line 63 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1mADO2NlX3ZgioqSd:-MzGV2IOBpUMLzdwDDEQ:b-xlzl5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3b1fed88cca24493a780140a530b9bf020f912b/pkg/util/httputil/handlers.go#L63)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: stylewise, we would typically return here and have the else clause on its own as the last line in the function.
</blockquote></details>

ezpz, can do!

**Review by sjbarag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78460)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @irfansharif)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-MzHTEb88Y5-Bc8A5SSV:-MzHTEb88Y5-Bc8A5SSW:b-vol2l9) (no related file):*
Thanks for the review! :smile: 

---
___
*[pkg/util/httputil/etag_handler_test.go, line 54 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1meBhEJtBMyMk_a6Q:-MzHTDnH4HmquseJUFbi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3b1fed88cca24493a780140a530b9bf020f912b/pkg/util/httputil/etag_handler_test.go#L54)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

as discussed on Slack, probably need to close the server.
</blockquote></details>

Done.

**Review by dhartunian — APPROVED** [review]
> Reviewed 9 of 9 files at r9, 7 of 7 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78460)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @irfansharif, and @sjbarag)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-MzHTEb88Y5-Bc8A5SSV:-MzHYRK-Ajoyr-VNPRI8:bg7igkv) (no related file):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

Thanks for the review! :smile: 
</blockquote></details>

:lgtm: 

---
___
*[-- commits, line 32 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-MzHYKVyBAmQw03R5Zpc):*
merges in CRDB do not squash so would you mind pushing the fixup into the relevant commit?
___
*[pkg/ui/buildutil/hash_files.go, line 56 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78460#-Mz1jnL-BDzzQakP79d4:-MzHXt8ZED2VV8fpJYBt:b-dmzd9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/431671b927001264a34742ceb4bf309544f9782b/pkg/ui/buildutil/hash_files.go#L56)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

`fs.WalkDir` passes a relative path to its callback as the `d` param, so nested assets would end up in `fileHashes` as `foo/bar/baz.js`.  The path provided in a request's `r.URL.Path` has a leading `/` though!  It seemed reasonable to include it here, but there's definitely some incongruity.

Think I should I rename `fileHashes` to `pathHashes` and keep the leading `/` here?  Or should I maybe remove the leading `/` from the `r.URL.Path` in the `ETag` handler?
</blockquote></details>

meh, it's fine to keep as is. the usage is so minimal.

**Comment by sjbarag** [comment]
Hoping to get a backport filed tonight, so let's bump the priority.

bors p=1

**Pull request merged**